File: src/algorithms/pathfinding/shortest-path/a-star/__tests__/a-star-grid_test.py
import importlib
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))

a_star_grid_mod = importlib.import_module("a-star-grid")
a_star_grid = a_star_grid_mod.a_star_grid


def make_grid(rows, cols):
    return [[{"type": "empty"} for _ in range(cols)] for _ in range(rows)]


def test_finds_path():
    grid = make_grid(5, 5)
    grid[0][0]["type"] = "start"
    grid[4][4]["type"] = "end"
    result = a_star_grid(grid, (0, 0), (4, 4))
    assert len(result["path"]) > 0


def test_shortest_path_length():
    grid = make_grid(5, 5)
    grid[0][0]["type"] = "start"
    grid[4][4]["type"] = "end"
    result = a_star_grid(grid, (0, 0), (4, 4))
    assert len(result["path"]) == 9


def test_path_empty_when_blocked():
    grid = make_grid(3, 3)
    grid[0][0]["type"] = "start"
    grid[2][2]["type"] = "end"
    for row in range(3):
        grid[row][1]["type"] = "wall"
    result = a_star_grid(grid, (0, 0), (2, 2))
    assert len(result["path"]) == 0


def test_navigates_around_wall():
    grid = make_grid(5, 5)
    grid[0][0]["type"] = "start"
    grid[4][4]["type"] = "end"
    for row in range(4):
        grid[row][2]["type"] = "wall"
    result = a_star_grid(grid, (0, 0), (4, 4))
    assert len(result["path"]) > 0


def test_adjacent_cells():
    grid = make_grid(3, 3)
    grid[0][0]["type"] = "start"
    grid[0][1]["type"] = "end"
    result = a_star_grid(grid, (0, 0), (0, 1))
    assert len(result["path"]) == 2


def test_tracks_visited():
    grid = make_grid(5, 5)
    grid[0][0]["type"] = "start"
    grid[4][4]["type"] = "end"
    result = a_star_grid(grid, (0, 0), (4, 4))
    assert len(result["visited"]) > 0


if __name__ == "__main__":
    test_finds_path()
    test_shortest_path_length()
    test_path_empty_when_blocked()
    test_navigates_around_wall()
    test_adjacent_cells()
    test_tracks_visited()
    print("All tests passed!")

File: src/algorithms/pathfinding/shortest-path/a-star/a-star-grid_test.go
package astar

import "testing"

func makeGrid(rows, cols int) [][]GridCell {
	grid := make([][]GridCell, rows)
	for row := range grid {
		grid[row] = make([]GridCell, cols)
		for col := range grid[row] {
			grid[row][col] = GridCell{Row: row, Col: col, CellType: CellEmpty, State: "default"}
		}
	}
	return grid
}

func TestFindsPath(t *testing.T) {
	grid := makeGrid(5, 5)
	grid[0][0].CellType = CellStart
	grid[4][4].CellType = CellEnd
	result := AStarGrid(grid, [2]int{0, 0}, [2]int{4, 4})
	if len(result.Path) == 0 {
		t.Error("expected path to be found")
	}
}

func TestShortestPathLength(t *testing.T) {
	grid := makeGrid(5, 5)
	grid[0][0].CellType = CellStart
	grid[4][4].CellType = CellEnd
	result := AStarGrid(grid, [2]int{0, 0}, [2]int{4, 4})
	if len(result.Path) != 9 {
		t.Errorf("expected path length 9, got %d", len(result.Path))
	}
}

func TestPathEmptyWhenBlocked(t *testing.T) {
	grid := makeGrid(3, 3)
	grid[0][0].CellType = CellStart
	grid[2][2].CellType = CellEnd
	for row := 0; row < 3; row++ {
		grid[row][1].CellType = CellWall
	}
	result := AStarGrid(grid, [2]int{0, 0}, [2]int{2, 2})
	if len(result.Path) != 0 {
		t.Error("expected empty path when blocked")
	}
}

func TestNavigatesAroundWall(t *testing.T) {
	grid := makeGrid(5, 5)
	grid[0][0].CellType = CellStart
	grid[4][4].CellType = CellEnd
	for row := 0; row < 4; row++ {
		grid[row][2].CellType = CellWall
	}
	result := AStarGrid(grid, [2]int{0, 0}, [2]int{4, 4})
	if len(result.Path) == 0 {
		t.Error("expected path around wall")
	}
}

func TestAdjacentCells(t *testing.T) {
	grid := makeGrid(3, 3)
	grid[0][0].CellType = CellStart
	grid[0][1].CellType = CellEnd
	result := AStarGrid(grid, [2]int{0, 0}, [2]int{0, 1})
	if len(result.Path) != 2 {
		t.Errorf("expected path length 2, got %d", len(result.Path))
	}
}

func TestTracksVisited(t *testing.T) {
	grid := makeGrid(5, 5)
	grid[0][0].CellType = CellStart
	grid[4][4].CellType = CellEnd
	result := AStarGrid(grid, [2]int{0, 0}, [2]int{4, 4})
	if len(result.Visited) == 0 {
		t.Error("expected visited cells to be tracked")
	}
}
